File: 1. RPN Calculator/rpn-calculator.pl
# Programming Praxis
# RPN Calculator
# http://programmingpraxis.com/2009/02/19/rpn-calculator/
# Without any error handling.
#
# Sergey Dymchenko <kit1980@gmail.com>
#
# Language: Perl
# Tested with perl 5.10.1
# Usage: perl rpn-calculator.pl

use strict;
use warnings;

use Scalar::Util qw(looks_like_number);

package RPN;

sub new {
    my $class = shift;
    my $self = {
        stack => [],
    };
 
    bless $self, $class;
    return $self;
}

sub process {
    my $self = shift;
    my $expression_string = shift;
    my @tokens = split(' ', $expression_string);
    foreach my $token (@tokens) {
        if (Scalar::Util::looks_like_number($token)) {
            # this is a number, push to the stack
            push @{$self->{stack}}, $token;
        } else {
            # an operation, do it and push the result to the stack
            my $y = pop @{$self->{stack}}; # note y is on top of the stack
            my $x = pop @{$self->{stack}};
            my $result;

            $result = $x + $y if $token eq "+";
            $result = $x - $y if $token eq "-";
            $result = $x * $y if $token eq "*";
            $result = $x / $y if $token eq "/";

            push @{$self->{stack}}, $result;
        }
    }
}

sub result {
    my $self = shift;
    return $self->{stack}[-1];
}

package main;

my $calc = RPN->new();
while (my $expression_string  = <STDIN>) {
    $calc->process($expression_string);
    print $calc->result(), "\n";
}

File: 1. RPN Calculator/rpn-calculator.py
# Programming Praxis
# RPN Calculator
# http://programmingpraxis.com/2009/02/19/rpn-calculator/
# Without any error handling.
#
# Sergey Dymchenko <kit1980@gmail.com>
#
# Language: Python
# Tested with Python 2.6.5
# Usage: python rpn-calculator.py


class RPNCalc(object):

    def __init__(self):
        self.stack = []
        self.operations = {
            "+": lambda x, y: x + y,
            "-": lambda x, y: x - y,
            "*": lambda x, y: x * y,
            "/": lambda x, y: x / y,
            }

    def process(self, expression_string):
        tokens = expression_string.split()
        for token in tokens:
            try:
                # this is a number, push to the stack
                n = float(token)
                self.stack.append(n)
            except ValueError:
                # an operation, do it and push the result to the stack
                y = self.stack.pop() # note y is on the top of the stack
                x = self.stack.pop()
                result = self.operations[token](x, y)
                self.stack.append(result)

    def result(self):
        return self.stack[-1]


def main():
    calc = RPNCalc()
    while True:
        try:
            expression_string = raw_input()
            calc.process(expression_string)
            print calc.result()
        except EOFError:
            break

if __name__ == "__main__":
    main()
